cmd/kube-apiserver/app/options/options_test.go

```diff
@@ -125,6 +126,7 @@ func TestAddFlags(t *testing.T) {
 		GenericServerRunOptions: &apiserveroptions.ServerRunOptions{
 			AdvertiseAddress:            net.ParseIP("192.168.10.10"),
 			CorsAllowedOriginList:       []string{"10.10.10.100", "10.10.10.200"},
+			MaxRequestsPerUserInFlight:  100,
 			MaxRequestsInFlight:         400,
```

Comment:
Just looking at this file, I'm not convinced this design is going to work. I think it will be difficult to impossible to tune these two flags. Either it won't actually offer than much protection (100/400 means it only takes 4 misbehaving users to DoS the server) or it will block actually important users (kubelets, controllers/schedulers). There is probably a narrow window of useful values and it probably varies a lot with cluster size and workload.

Sorry, but I would like to read a design doc that addresses this concern before we proceed with this. The thing that I would find convincing is a graph(s) or chart(s) (backed by actual cluster request qps data) demonstrating that the window of reasonable settings is much larger than I assume, and valid for a large range of cluster sizes and workloads. Without that, my first impression is that this is just making the life of a cluster operator really hard.

@wojtek-t might have thoughts.

Reply:
Sorry for the delayed response.

> Can the attacker (someone deliberately DOSing) set any part of the user string?
It seems, that attacker can't set nor username nor group, unless he already has User Impersonation privileges.

I agree with @lavalamp's remarks, that "max-requests-per-user-inflight" option is suitable for very few use cases.
Accordign to this, what do you think about folowing feature:

Implement `readonly_limit`, `readonly_request` and `mutating_limit`, `mutating_request` by group, for readonly & mutating requests accordingly:
`limit` just limits total inflight requests across particular group.
`request` - inflight requests, reserved for particular group. E.g.: We have `maxInflightRequest` == 400, `readonly_request` specified for `system:masters` is 5, for `system:node` is 10, the maximum total inflight requests of users from other groups is 385. In other words, members of `system:masters` in any time can to have at least 5 inflight requests.

Limits and requests provided via config yaml file with following format:
```
---
%GROUP_NAME_1%:
  readonly_limit: %READONLY_LIMIT_1%
  readonly_request: %READONLY_REQUEST_1%
  mutating_limit: %MUTATING_LIMIT_1%
  mutating_request: %MUTATING_REQUEST_1%
%GROUP_NAME_2%:
  readonly_limit: %READONLY_LIMIT_2%
  readonly_request: %READONLY_REQUEST_2%
  mutating_limit: %MUTATING_LIMIT_2%
  mutating_request: %MUTATING_REQUEST_2%
```


Algorithm to accept/reject readonly request:

0. pick group with max `readonly_request` across user's group as `requestGroup`, pick group with max `readonly_limit` across user's group as `limitGroup`
1. if total inflight requests of `requestGroup` > `limitGroup.readonly_limit`, REJECT inbound request
2. else, if total inflight requests of `requestGroup` < `requestGroup.readonly_request`, ALLOW inbound request
3. else, if (sum of `readonly_request` of all groups + all readonly requests, exeedes appropriate `requestGroup.readonly_request`) < `max-requests-inflight`, ALLOW inbound request
4. else, REJECT  inbound request

For mutating requests algorithm is the same.

What do you think about it? Maybe, better to open proposal?

Reply:
There is one thing that I would like to see before that.

Which is somehow refining "readonly_limit".
Current getting a single pod is treated as one request the same way as "listing all pods" - that is obviously wrong, because the latter is waaay more expensive.

What I would like to see is making that more realistic. In particular, if there 1000 pods in the system, list all pods request should consume 1000 (or maybe not 1000, because there is constant per-object overhead, but something proportional to 1000) tokens, while getting a single pod should consume just one token.

I don't have any precise suggestions how to do that exactly, but I think that's one of the problems we need to solve quickly - otherwise readonly limits are pretty much useless for protecting us from being overloaded.